backend/service/chaos/experimentation/terminator/terminator.go

```diff
@@ -0,0 +1,156 @@
+package terminator
+
+import (
+	"context"
+	"sync/atomic"
+	"time"
+
+	"github.com/golang/protobuf/ptypes"
+	"github.com/uber-go/tally"
+	"go.uber.org/zap"
+
+	experimentationv1 "github.com/lyft/clutch/backend/api/chaos/experimentation/v1"
+	"github.com/lyft/clutch/backend/service/chaos/experimentation/experimentstore"
+)
+
+type TerminationCriteria interface {
+	ShouldTerminate(experimentStarted time.Time, config interface{}) error
+}
+
+type Monitor interface {
+	Run(ctx context.Context)
+}
+
+// TODO(snowp): Remove this once we have a proper service object that we can create.
+func NewTestMonitor(store experimentstore.Storer, enabledConfigTypes []string, criterias []TerminationCriteria, log *zap.SugaredLogger, stats tally.Scope) Monitor {
+	return &monitor{
+		store:                      store,
+		enabledConfigTypes:         enabledConfigTypes,
+		criterias:                  criterias,
+		outerLoopInterval:          1,
+		perExperimentCheckInterval: 1,
+		log:                        log,
+		activeMonitoringRoutines:   trackingGauge{gauge: stats.Gauge("active_monitoring_routines")},
+		terminationCount:           stats.Counter("terminations"),
+	}
+}
+
+type monitor struct {
+	store              experimentstore.Storer
+	enabledConfigTypes []string
+	criterias          []TerminationCriteria
+
+	outerLoopInterval          time.Duration
+	perExperimentCheckInterval time.Duration
+
+	log *zap.SugaredLogger
+
+	activeMonitoringRoutines trackingGauge
+	terminationCount         tally.Counter
+}
+
+func (m *monitor) Run(ctx context.Context) {
+	// Start a single goroutine that monitors all the active experiments at a fixed interval. Whenever a new (new to this goroutine)
+	// experiment is found, open up a goroutine that periodically evaluates all the termination criteria for the experiment.
+	//
+	// This approach ensures provides a steady DB pressure (mostly outer loop, some from triggering termination) and relatively high
+	// fairness, as checking the termination conditions for one experiment should not be delaying the checks for another experiment
+	// unless we're under very heavy load.
+	go func() {
+		trackedExperiments := map[uint64]context.CancelFunc{}
+		ticker := time.NewTicker(m.outerLoopInterval)
+
+		for {
+			select {
+			case <-ctx.Done():
+				return
+			case <-ticker.C:
+				es, err := m.store.GetExperiments(context.Background(), m.enabledConfigTypes, experimentationv1.GetExperimentsRequest_STATUS_RUNNING)
```

Comment:
Did we update the implementation of GetExperiments so that it supports array of config types now? I cannot find this change. Or is the plan to update it in the following PRs?

Reply:
The current PR is missing some changes for that, I was gonna do it in here

Right now I think we would blow up at startup since the real store doesn't satisfy the required interface